File: Go/matrices/lowerTriangularMatrix.go
package matrices

import (
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
)

// Lower-triangular matrix structure.
type LowerTriangularMatrix struct {
	// Size.
	Size int
	// Total number of all non-zero elements.
	ElemCount int
	// Matrix itself.
	Elems []float64
}

// Creating an instance of a lower-triangular matrix and returning a pointer.
func NewLowerTriangularMatrix(size int) *LowerTriangularMatrix {
	ltm := new(LowerTriangularMatrix)
	ltm.Size = size
	ltm.ElemCount = size * (size + 1) / 2
	ltm.Elems = make([]float64, ltm.ElemCount)
	return ltm
}

// File input.
func (ltm *LowerTriangularMatrix) In(line string) {
	strs := strings.Split(line, " ")
	for i := 0; i < ltm.ElemCount; i++ {
		ltm.Elems[i], _ = strconv.ParseFloat(strs[i], 64)
	}
}

// Random input.
func (ltm *LowerTriangularMatrix) RandomIn() {
	for i := 0; i < ltm.ElemCount; i++ {
		ltm.Elems[i] = float64(rand.Int31()%100) * rand.Float64()
	}
}

// File output.
func (ltm *LowerTriangularMatrix) Out(f *os.File) {
	f.WriteString("Size: ")
	f.WriteString(fmt.Sprintf("%d\n", ltm.Size))
	f.WriteString("Average: ")
	f.WriteString(fmt.Sprintf("%f\n", ltm.GetAverage()))
	f.WriteString("Type: Lower-Triangular Matrix\n")

	for i := 0; i < ltm.Size; i++ {
		for j := 0; j < ltm.Size; j++ {
			// Only elements which are placed lower or on the diagonal can be non-zero.
			if i >= j {
				f.WriteString(fmt.Sprintf("%f ", ltm.Elems[i]))
			} else {
				f.WriteString("0 ")
			}
		}
		f.WriteString("\n")
	}
}

// Getting average of all the elements.
func (ltm *LowerTriangularMatrix) GetAverage() float64 {
	// Total sum of all the elements.
	var sum float64 = 0

	for i := 0; i < ltm.ElemCount; i++ {
		sum += ltm.Elems[i]
	}

	return sum / float64(ltm.Size*ltm.Size)
}

File: Python/matrices/lowerTriangularMatrix.py
from matrices.baseMatrix import BaseMatrix
import random


# Lower-triangular matrix class.
class LowerTriangularMatrix(BaseMatrix):
    # Lower-triangular matrix size.
    __size__: int
    # Total number of non-zero elements.
    __elems_count__: int
    # Lower-triangular matrix itself.
    __elems__: list

    # Lower-triangular matrix constructor.
    def __init__(self, size: int):
        super().__init__(size)
        self.__elems_count__ = self.__size__ * (self.__size__ + 1) // 2
        self.__elems__ = []

    # Reading from file.
    def read(self, line: str):
        # Filling lower-triangular matrix with numbers.
        self.__elems__ = list(map(float, line.split()))

    # Creating randomly.
    def random_read(self):
        self.__size__ = random.randint(0, 100) + 1
        self.__elems_count__ = self.__size__ * (self.__size__ + 1) // 2
        self.__elems__ = []

        # Filling lower-triangular matrix with numbers.
        for i in range(0, self.__elems_count__):
            self.__elems__.append(random.random() * random.randint(0, 1000))

    # Outputting.
    def out(self):
        ans = 'Size: ' + str(self.__size__) + '\n'
        ans += 'Average: ' + str(self.get_average()) + '\n'
        ans += 'Lower-triangular matrix:\n'
        counter = 0

        for i in range(0, self.__size__):
            for j in range(0, self.__size__):
                if i < j:
                    ans += '0 '
                else:
                    ans += str(self.__elems__[counter]) + ' '
                    counter += 1
            ans += '\n'

        return ans

    # Getting average.
    def get_average(self) -> float:
        # Total sum of elements.
        summ = 0

        for i in range(0, self.__elems_count__):
            summ = summ + self.__elems__[i]

        return summ / self.__size__ / self.__size__
